package/sbp_fileio_daemon/sbp_fileio_daemon/src/sbp_fileio.c

```diff
@@ -53,32 +103,406 @@ static void read_dir_cb(u16 sender_id, u8 len, u8 msg[], void *context);
+static void purge_fd_cache(pk_loop_t *loop, void *handle, int status, void *context);
+static void flush_cached_fd(const char *path, const char *mode);
+static fd_cache_result_t open_from_cache(const char *path,
+                                         const char *mode,
+                                         int oflag,
+                                         mode_t perm);
+
+static bool test_control_dir(const char *name);
+
+static fd_cache_t fd_cache[FD_CACHE_COUNT] = {[0 ...(FD_CACHE_COUNT - 1)] = {NULL, "", "", 0, 0}};
+
 path_validator_t *g_pv_ctx;
 
+fd_cache_result_t empty_cache_result = (fd_cache_result_t){
+  .fp = NULL,
+  .cached = false,
+  .offset = NULL,
+};
+
+fd_cache_t empty_cache_entry = (fd_cache_t){
+  .fp = NULL,
+  .path = "",
+  .mode = "",
+  .opened_at = -1,
+  .offset = 0,
+};
+
+static off_t read_offset(fd_cache_result_t r)
+{
+  if (r.offset == NULL) return -1;
+  return (off_t)*r.offset;
+}
+
+static void update_offset(fd_cache_result_t r, off_t offset, int op)
+{
+  if (r.offset == NULL) return;
+
+  if (op == OP_INCREMENT)
+    (*r.offset) += offset;
+  else if (op == OP_ASSIGN)
+    (*r.offset) = offset;
+  else
+    PK_LOG_ANNO(LOG_ERR, "invalid operation: %d", op);
+}
+
+static void flush_fd_cache()
+{
+  for (size_t idx = 0; idx < FD_CACHE_COUNT; idx++) {
+
+    if (fd_cache[idx].fp == NULL) continue;
+
+    fclose(fd_cache[idx].fp);
+    fd_cache[idx] = empty_cache_entry;
+  }
+}
+
+static void purge_fd_cache(pk_loop_t *loop, void *handle, int status, void *context)
+{
+  (void)loop;
+  (void)status;
+  (void)handle;
+  (void)context;
+
+  time_t now = time(NULL);
+
+  for (size_t idx = 0; idx < FD_CACHE_COUNT; idx++) {
+    if (fd_cache[idx].fp == NULL) continue;
+    if ((now - fd_cache[idx].opened_at) >= CACHE_CLOSE_AGE) {
+      fclose(fd_cache[idx].fp);
+      fd_cache[idx] = empty_cache_entry;
+    }
+  }
+
+  pk_metrics_flush(MR);
+
+  pk_metrics_reset(MR, MI.read_bytes);
+  pk_metrics_reset(MR, MI.write_bytes);
+
+  pk_loop_timer_reset(cleanup_timer_handle);
+}
+
+static void return_to_cache(fd_cache_result_t *cache_result)
+{
+  if (!cache_result->cached && cache_result->fp != NULL) {
+    fclose(cache_result->fp);
+    return;
+  }
+}
+
+static void flush_cached_fd(const char *path, const char *mode)
+{
+  for (size_t idx = 0; idx < FD_CACHE_COUNT; idx++) {
+    if (strncmp(fd_cache[idx].path, path, sizeof(fd_cache[idx].path)) == 0
+        && strncmp(fd_cache[idx].mode, mode, sizeof(fd_cache[idx].mode)) == 0) {
+      FIO_LOG_DEBUG("Flushing cached write fp (index %d): %p, filename: %s, mode: %s, offset: %d",
+                    idx,
+                    fd_cache[idx].fp,
+                    fd_cache[idx].path,
+                    fd_cache[idx].mode,
+                    fd_cache[idx].offset);
+      fclose(fd_cache[idx].fp);
+      fd_cache[idx] = empty_cache_entry;
+    }
+  }
+}
+
+static fd_cache_result_t open_from_cache(const char *path, const char *mode, int oflag, mode_t perm)
```

Comment:
Seeing that there is `no_cache` variable leads me to thinking is this `open_from_cache` name a bit misleading in some cases?

Reply:
Changed naming to make the code less misleading